Add RELEASE_ASSERT() type check to downcast<>() on Intel

cdumez opened the pull request:
#### 105163950fa364aba7685c4077061930da3251ec
<pre>
Add RELEASE_ASSERT() type check to downcast&lt;&gt;() on Intel
<a href="https://bugs.webkit.org/show_bug.cgi?id=271470">https://bugs.webkit.org/show_bug.cgi?id=271470</a>

Reviewed by Darin Adler.

Add RELEASE_ASSERT() type check to downcast&lt;&gt;() on Intel, to match what we
already do on ARM. This tested as performance neutral on the benchmarks we
track.

* Source/WTF/wtf/TypeCasts.h:
(WTF::downcast):

Canonical link: <a href="https://commits.webkit.org/276552@main">https://commits.webkit.org/276552@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b338cdbf27f20474202880e5f6276594f96108d0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/44955 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24064 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47454 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/47613 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40962 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28134 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21462 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36908 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45533 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21126 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/38716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/17994 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/18550 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/39849 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3003 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/38155 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/41226 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/40151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49285 "Built successfully") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/44412 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/19927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16479 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/43907 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21243 "Built successfully") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/42676 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/21584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/51583 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6252 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20922 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/10476 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/TypeCasts.h:
```diff
@@ -127,13 +121,7 @@ inline match_constness_t<Source, Target>* downcast(Source* source)
-    // FIXME: This is too expensive to enable on x86 for now but we should try and
-    // enable the RELEASE_ASSERT() on all architectures.
-#if CPU(ARM64)
     RELEASE_ASSERT(!source || is<Target>(*source));
```
Can’t this be written as follows?
```suggestion
    RELEASE_ASSERT(is<Target>(source));
```

annevk replied:
That doesn't allow source to be null. The status quo does.

cdumez replied:
Yes, I'll update. Thanks.

cdumez replied:
Hmm, actually this wouldn't work.
`RELEASE_ASSERT(is<Target>(source));` is equivalent to `RELEASE_ASSERT(source && is<Target>(*source));`, not `RELEASE_ASSERT(!source || is<Target>(*source));`

annevk replied:
Oh good. You had me questioning my understanding of things for a moment there.

darinadler replied:
> That doesn't allow source to be null. The status quo does.

Well spotted. I’m glad I worded my suggestion as a question. Less embarrassing that I got it wrong.

cdumez committed:
Add RELEASE_ASSERT() type check to downcast<>() on Intel
https://bugs.webkit.org/show_bug.cgi?id=271470

Reviewed by Darin Adler.

Add RELEASE_ASSERT() type check to downcast<>() on Intel, to match what we
already do on ARM. This tested as performance neutral on the benchmarks we
track.

* Source/WTF/wtf/TypeCasts.h:
(WTF::downcast):

Canonical link: https://commits.webkit.org/276552@main
